docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,123 @@
+
+- reference: "I want to use this entry as-is, including its configuration data"
+- template: "I want to create items from the specified template"
+- provision: "I want the creation to be goverened by some other entity that implements the 'service broker' HTTP interface"
+
+We see the type as an arbitrary `string`; one or more controllers could run to process each claim type, performing whatever logic is appropriate to fulfill the specific type in question.
+
+### Viewing a catalog
+
+Users should be able to list the entries in a service catalog. Users should be able to select an entry and "consume" it. We call this consumption "claiming" a service from a service catalog.
+
+### Claiming a service
+
+When you want to use an entry from the service catalog, you create a `ServiceClaim` that references the desired entry. A controller processes new claims for admission. This determines if the user who created the claim is allowed to consume the entry from the catalog. This decision can be flexible: it could be automated based on policy, or it could support manual intervention and workflow.
+
+Once the claim has been admitted, a controller performs the provisioning process based on the service catalog entry's type.
+
+For a reference claim, a new service is created in the user's namespace. The new service somehow needs to be a CNAME to the source. Additionally, for each configuration resource referenced by the source service, the controller creates a clone of it in the user's namespace.
+
+For a template claim, the template is processed and its items are created in the user's namespace (see [OpenShift templates](https://docs.openshift.org/latest/dev_guide/templates.html) for more details on templates).
+
+For a provision claim, the controller sends requests to the specified "service broker" endpoint to perform whatever provisioning actions the broker is coded to do.
+
+Additional fulfillment types are possible as long as there is a controller that is handling them.
+
+TODO poking holes in cluster firewall for cross-namespace connections
+
+## Linking services
+
+Claiming a service catalog entry only creates resources in the user's namespace. If all you need is a service and its DNS entry, this may be sufficient for your pods to function. But if you need the configuration data injected into your pod, it would be nice to make that easier to do.
+
+We want to add the ability to link a service to a deployable resource such as a Deployment. This could look something like:
+
+    kubectl link postgresql deployment/web
+    service "postgresql" linked to deployemnt "web"
+    configdata "postgresql-options" linked to deployment "web" as a volume
+    secret "postgresql-credentials" linked to deployment "web" as a volume
+
+This command would automatically inject the ConfigData and Secret objects as volumes into the Deployment. This could be flexible as well, allowing you instead to expose these items as environment variables.
```

Comment:
Yes, magic is bad.  I should be able to inspect the pod template and
identify that I am consuming a secret, config data, and probably dependent
on a service.  The mechanics of how that linkage happens is a pretty high
bar (in terms of number of concepts known).  The idea of the link is to
reduce the integration to a few well chosen defaults and offer a "good by
convention path".

Regarding shopping for a database, he primary use of infrastructure
abstraction is to reduce the operational cost of taking advantage of things
an operational team supports.  Since there is no database everyone uses,
but almost everything needs some form of persistent store, you look for
opportunities for administrators / operators / db experts to reduce the
friction around instantiating / running those persistent stores.  The easy
win is usually the simplest - provide db images / pod templates that "just
work".  The next step is harder - making them operationally reliable.
That's typically where an ops team decouples them from their consumers -
providing an interface / web page / API where a client says "hey, can I
have a production db"?  That today is usually a ticketed process (hey, can
you spin me one?  Sure, check back in 3 weeks) - our goal should be to
provide the glue that makes the request / allocation trend to zero in
incremental cost, but allow the databases to be maintained by the people
with the necessary magic spells.


On Nov 23, 2015, at 5:09 PM, Tim Hockin <notifications@github.com> wrote:

In docs/proposals/service-catalog.md
<https://github.com/kubernetes/kubernetes/pull/17543#discussion_r45668308>:

> +Additional fulfillment types are possible as long as there is a controller that is handling them.
> +
> +TODO poking holes in cluster firewall for cross-namespace connections
> +
> +## Linking services
> +
> +Claiming a service catalog entry only creates resources in the user's namespace. If all you need is a service and its DNS entry, this may be sufficient for your pods to function. But if you need the configuration data injected into your pod, it would be nice to make that easier to do.
> +
> +We want to add the ability to link a service to a deployable resource such as a Deployment. This could look something like:
> +
> +    kubectl link postgresql deployment/web
> +    service "postgresql" linked to deployemnt "web"
> +    configdata "postgresql-options" linked to deployment "web" as a volume
> +    secret "postgresql-credentials" linked to deployment "web" as a volume
> +
> +This command would automatically inject the ConfigData and Secret objects as volumes into the Deployment. This could be flexible as well, allowing you instead to expose these items as environment variables.

I lost track of this argument somewhere.

If this is done after the fact, won't my deployment already be in progress
by the time I link it? That sounds terrible.

As for shopping for a database, that's such a foreign idea to me I don't
know where to start. But since you're describing claims... user goes
shopping, puts a claim on a DB service, gets the associated Secret object,
then creates their deployment. It really needs to be obvious from the
deployment what linkages are happening. Magic is bad.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17543/files#r45668308>.